File: js/algorithms/quicksort.js
var utils = require('./utils/array'),
    swap = utils.swap;

function sort(array, start, end) {
    start = typeof start !== 'undefined' ? start : 0;
    end = typeof end !== 'undefined' ? end : (array.length - 1);
    
    if (start >= end) {
        return;
    }
    
    var pivot = (end + start) >> 1,
        value = array[pivot],
        position = start,
        i;
    
    swap(array, pivot, end);
    
    for (i = start; i < end; i++) {
        
        if (array[i] < value) {
            swap(array, i, position);
            position += 1;
        }
    }

    swap(array, end, position);

    sort(array, start, position);
    sort(array, position + 1, end);

}

exports.name = 'quicksort';
exports.sort = sort;

File: c++/algorithms/quicksort/quicksort.cpp
#include "quicksort.h"

void QuickSort::sort(int *array, int length) {
	this->sort(array, 0, length - 1);
}

void QuickSort::sort(int *array, int start, int end) {

	if (start >= end) {
		return;
	}


	int i, 
		dividerLocation = (start + end) >> 1, 
		dividerValue = array[dividerLocation],
		newDividerLocation = start;

	swap(array, dividerLocation, end);

	for (i = start; i < end; i++) {
		if (array[i] < dividerValue) {
			swap(array, i, newDividerLocation);
			newDividerLocation++;
		}
	}

	swap(array, end, newDividerLocation);

	this->sort(array, start, newDividerLocation);
	this->sort(array, newDividerLocation + 1, end);
}